packages/protocol/contracts/layer1/verifiers/SgxVerifierBase.sol

```diff
@@ -0,0 +1,174 @@
+
+    error SGX_ALREADY_ATTESTED();
+    error SGX_INVALID_ATTESTATION();
+    error SGX_INVALID_INSTANCE();
+    error SGX_RA_NOT_SUPPORTED();
+
+    /// @notice Register an SGX instance after the attestation is verified
+    /// @param _attestation The parsed attestation quote.
+    /// @return The respective instanceId
+    function registerInstance(V3Struct.ParsedV3QuoteStruct calldata _attestation)
+        external
+        returns (uint256)
+    {
+        address automataDcapAttestation = resolve(LibStrings.B_AUTOMATA_DCAP_ATTESTATION, true);
+
+        if (automataDcapAttestation == address(0)) {
+            revert SGX_RA_NOT_SUPPORTED();
+        }
+
+        (bool verified,) = IAttestation(automataDcapAttestation).verifyParsedQuote(_attestation);
+
+        if (!verified) revert SGX_INVALID_ATTESTATION();
+
+        address[] memory addresses = new address[](1);
+        addresses[0] = address(bytes20(_attestation.localEnclaveReport.reportData));
+
+        return _addInstances(addresses, false)[0];
+    }
+
+    /// @notice Adds trusted SGX instances to the registry.
+    /// @param _instances The address array of trusted SGX instances.
+    /// @return The respective instanceId array per addresses.
+    function addInstances(address[] calldata _instances)
+        external
+        onlyOwner
+        returns (uint256[] memory)
+    {
+        return _addInstances(_instances, true);
+    }
+
+    /// @notice Deletes SGX instances from the registry.
+    /// @param _ids The ids array of SGX instances.
+    function deleteInstances(uint256[] calldata _ids)
+        external
+        onlyFromOwnerOrNamed(LibStrings.B_SGX_WATCHDOG)
+    {
+        for (uint256 i; i < _ids.length; ++i) {
+            uint256 idx = _ids[i];
+
+            if (instances[idx].addr == address(0)) revert SGX_INVALID_INSTANCE();
+
+            emit InstanceDeleted(idx, instances[idx].addr);
+
+            delete instances[idx];
+        }
+    }
+
+    function _addInstances(
+        address[] memory _instances,
+        bool instantValid
+    )
+        internal
+        returns (uint256[] memory ids)
+    {
+        ids = new uint256[](_instances.length);
+
+        uint64 validSince = uint64(block.timestamp);
+
+        if (!instantValid) {
+            validSince += INSTANCE_VALIDITY_DELAY;
+        }
+
+        for (uint256 i; i < _instances.length; ++i) {
+            if (addressRegistered[_instances[i]]) revert SGX_ALREADY_ATTESTED();
+
+            addressRegistered[_instances[i]] = true;
+
+            if (_instances[i] == address(0)) revert SGX_INVALID_INSTANCE();
+
+            instances[nextInstanceId] = Instance(_instances[i], validSince);
+            ids[i] = nextInstanceId;
+
+            emit InstanceAdded(nextInstanceId, _instances[i], address(0), validSince);
+
+            ++nextInstanceId;
+        }
+    }
+
+    function _replaceInstance(uint256 id, address oldInstance, address newInstance) internal {
```

Comment:
Can we cache the instance in a local variable, we can minimize storage reads then here in `_replaceInstance`

Reply:
Sorry which read?